From 6f202c32f6cfabc8198653fe286f7ffb8320bfd4 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 26 Aug 2021 10:43:33 +0300
Subject: [PATCH] [SW-46364] ppboot: adding stack overflow monitor

The Stack monitor is checking the magic word that is stored at the
last address of the stack. If the value is not preserved during
the execution, an error bit is raised.

Change-Id: I074c23aaa554209a473e61eb15ff02b9328b08ca
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/app/ppboot/ppboot.c b/app/ppboot/ppboot.c
index 2093b53..dbc3cb0 100644
--- a/app/ppboot/ppboot.c
+++ b/app/ppboot/ppboot.c
@@ -233,11 +233,32 @@
 	wfe();
 }
 
+__weak void ppboot_init_stack_mon_dev_specific(void) {}
+__weak void ppboot_check_stack_mon_dev_specific(void) {}
+
+void ppboot_init_stack_monitor(void)
+{
+	__stack = STACK_CANARY_MAGIC_WORD;
+	ppboot_init_stack_mon_dev_specific();
+}
+
+void ppboot_check_stack_monitor(void)
+{
+	if (__stack != STACK_CANARY_MAGIC_WORD) {
+		preboot_puts("The stack has oveflowed\n");
+		preboot_puts_hex("Stack Canary value = ", __stack, false);
+		cpu_boot_err_clrset(CPU_BOOT_ERR0_BOOT_FW_CRIT_ERR, true);
+	}
+
+	ppboot_check_stack_mon_dev_specific();
+}
+
 void ppboot_main(void)
 {
 	u64 loadaddr;
 	bool ret;
 
+	ppboot_init_stack_monitor();
 	/* Initialize internal if variables */
 	cpu_boot_if_init();
 	hl_plat_early_init();
@@ -261,6 +282,7 @@
 
 	preboot_delay();
 	ret = ppboot_load_preboot_img(&loadaddr);
+	ppboot_check_stack_monitor();
 	if (ret)
 		image_start(loadaddr);
 	hl_enter_recovery();
diff --git a/arch/gaudi/arm/common.c b/arch/gaudi/arm/common.c
index 3283d1e..eb7a09c 100644
--- a/arch/gaudi/arm/common.c
+++ b/arch/gaudi/arm/common.c
@@ -356,6 +356,7 @@
 void secondary_main(void)
 {
 	preboot_pcie_monitoring();
+	ppboot_check_stack_mon_dev_specific();
 
 	/* Hang in WFE until further notice */
 	while (1)
@@ -623,3 +624,14 @@
 	 */
 	return (!rst_ctrl || wfe_stat);
 }
+
+void ppboot_init_stack_mon_dev_specific(void)
+{
+	__stack2 = STACK_CANARY_MAGIC_WORD;
+}
+
+void ppboot_check_stack_mon_dev_specific(void)
+{
+	if (__stack2 != STACK_CANARY_MAGIC_WORD)
+		cpu_boot_err_clrset(CPU_BOOT_ERR0_BOOT_FW_CRIT_ERR, true);
+}
diff --git a/arch/gaudi/arm/common.h b/arch/gaudi/arm/common.h
index 635921e..69c7fc6 100644
--- a/arch/gaudi/arm/common.h
+++ b/arch/gaudi/arm/common.h
@@ -55,6 +55,8 @@
 /* Only BP2 == 0x4 - means 8 blocks - up to 512KB */
 #define PROT_MASK			(READ_STATUS_BP2 | READ_STATUS_SRWD)
 
+extern u32 __stack2;		/* stack2 last cell value */
+
 extern void secondary_entry(void);
 
 int gaudi_hbm_init(void);
diff --git a/include/preboot.h b/include/preboot.h
index dcaa966..08ca78a 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -82,6 +82,9 @@
 
 extern uintptr_t _start;	/* Start of ppboot/preboot region */
 extern uintptr_t __stack_top;	/* End of ppboot/preboot region */
+extern u32 __stack;		/* stack last cell value */
+
+#define STACK_CANARY_MAGIC_WORD	0x8BADF00D
 
 extern struct hl_fw_info_compat fw_info;
 /*
@@ -210,6 +213,9 @@
 void preboot_engines_init(void);
 void preboot_clear_wd_gpio(void);
 
+void ppboot_init_stack_mon_dev_specific(void);
+void ppboot_check_stack_mon_dev_specific(void);
+
 void preboot_main(void);
 
 #endif /* __ASSEMBLY__ */
